postgres-appliance/major_upgrade/inplace_upgrade.py

```diff
@@ -0,0 +1,727 @@
+#!/usr/bin/env python
+import json
+import logging
+import os
+import psutil
+import psycopg2
+import shlex
+import shutil
+import subprocess
+import sys
+import time
+import yaml
+
+from collections import defaultdict
+from threading import Thread
+from multiprocessing.pool import ThreadPool
+
+logger = logging.getLogger(__name__)
+
+
+def patch_wale_prefix(value, new_version):
+    from spilo_commons import is_valid_pg_version
+
+    if '/spilo/' in value and '/wal/' in value:  # path crafted in the configure_spilo.py?
+        basename, old_version = os.path.split(value.rstrip('/'))
+        if is_valid_pg_version(old_version) and old_version != new_version:
+            return os.path.join(basename, new_version)
+    return value
+
+
+def update_configs(new_version):
+    from spilo_commons import append_extentions, get_bin_dir, get_patroni_config, write_file, write_patroni_config
+
+    config = get_patroni_config()
+
+    config['postgresql']['bin_dir'] = get_bin_dir(new_version)
+
+    version = float(new_version)
+    shared_preload_libraries = config['postgresql'].get('parameters', {}).get('shared_preload_libraries')
+    if shared_preload_libraries is not None:
+        config['postgresql']['parameters']['shared_preload_libraries'] =\
+                append_extentions(shared_preload_libraries, version)
+
+    extwlist_extensions = config['postgresql'].get('parameters', {}).get('extwlist.extensions')
+    if extwlist_extensions is not None:
+        config['postgresql']['parameters']['extwlist.extensions'] =\
+                append_extentions(extwlist_extensions, version, True)
+
+    write_patroni_config(config, True)
+
+    # update wal-e/wal-g envdir files
+    restore_command = shlex.split(config['postgresql'].get('recovery_conf', {}).get('restore_command', ''))
```

Comment:
will that wok for the upgrade from the version 12 to the 13 ?

Reply:
You probably think about `restore_command` converted into a GUC in postgres 12.
Patroni configs are backward compatible, i.e. internally it will "move" `postgresql.recovery_conf.restore_command` to `postgresql.parameters.restore_command`. In Spilo we are using the old format for convenience.